ENH Add libsvm-like calibration procedure to CalibratedClassifierCV

lucyleeow opened the pull request:
<!--
Thanks for contributing a pull request! Please ensure you have taken a look at
the contribution guidelines: https://github.com/scikit-learn/scikit-learn/blob/master/CONTRIBUTING.md#pull-request-checklist
-->

#### Reference Issues/PRs
closes #16145
Follows from PR #16167
The first step was to refactor `CalibratedClassifierCV`, which was done in #17803. This is a branch off of #17803. (will close #17803)


#### What does this implement/fix? Explain your changes.
Add `ensemble` option to `CalibratedClassifierCV`.
`ensemble=True` would be the default and would be the current implementation.
`ensemble=False` would be the same as the [libsvm implementation](https://github.com/scikit-learn/scikit-learn/blob/master/sklearn/svm/src/libsvm/svm.cpp#L2099) (used when `probability=True`):

When `cv` is not `'prefit'`:
- Concatenate predictions from each cv. Use all predictions to train calibrator.
- Classifiers fit after each cv would be [destroyed](https://github.com/scikit-learn/scikit-learn/blob/master/sklearn/svm/src/libsvm/svm.cpp#L2169)
- Use classifier fit on all data at prediction time

Note `ensemble=False` is the procedure described in the original Platt paper (see https://github.com/scikit-learn/scikit-learn/issues/16145#issue-551351752)

#### Any other comments?


<!--
Please be aware that we are a loose team of volunteers so patience is
necessary; assistance handling other issues is very welcome. We value
all user contributions, no matter how minor they are. If we are slow to
review, either the pull request needs some benchmarking, tinkering,
convincing, etc. or more likely the reviewers are simply busy. In either
case, we ask for your understanding during the review process.
For more information, see our FAQ on this topic:
http://scikit-learn.org/dev/faq.html#why-is-my-pull-request-not-getting-any-attention.

Thanks for contributing!
-->

lucyleeow committed:
wip

lucyleeow committed:
refactor

lucyleeow committed:
merge master

lucyleeow committed:
clean up, tests pass

lucyleeow committed:
remove debugging test

lucyleeow committed:
lint

lucyleeow committed:
suggestions

lucyleeow committed:
check class y

lucyleeow committed:
add att docstr

lucyleeow committed:
wip

lucyleeow committed:
first iter, tests pass

lucyleeow committed:
lint

lucyleeow commented:
@NicolasHug, this is probably not going to be a problem but I thought I would mention it. 

We use `cross_val_predict` to get the predictions, where if a training subset is missing 1 or more classes, it is replaced by 0 if `predict_proba` otherwise by ('we approximate negative infinity') 'the minimum finite float value for the dtype' for `decision_function`.
`CalibratedClassifierCV` uses 0 for both `predict_proba` and `decision_function`.

NicolasHug commented on sklearn/calibration.py:
```diff
@@ -236,28 +255,73 @@ def fit(self, X, y, sample_weight=None):
+                                       " method.")
+                df = cross_val_predict(base_estimator, X, y, cv=cv,
+                                       method=base_estimator_method)
+                if base_estimator_method == "decision_function":
+                    if df.ndim == 1:
+                        df = df[:, np.newaxis]
+                else:
+                    if len(self.label_encoder_.classes_) == 2:
+                        df = df[:, 1:]
```
The duplication is a bit unfortunate, there should be a way to merge this logic with that of `_get_predictions`

Maybe by extracting this into a `_reshape_df` utility?
Or maybe we can even let `_get_predictions` call `cross_val_predict` somehow, so that we can also merge the method checks above. 
Also is `df = df[:, 1:]` just equivalent to `df = df[:, 1]`?

lucyleeow replied:
> Or maybe we can even let `_get_predictions` call `cross_val_predict` somehow, so that we can also merge the method checks above.

This is what I was trying to do before - but `_get_predictions` is is used within `_fit_calibrator` which is within the cv for loop. This is why I wanted to put the cv for loop within `_get_predictions` but it got complicated.

I will add a `_reshape_df` function. Also I have a little dislike for the `df` naming as it is not always 'decision_function`. (I also first think df=dataframe but maybe that is just me)

lucyleeow replied:
Also I can't think of an elegant way to add the `_reshape_df` function as we get the predictions of the classifier in `_get_predictions` but we use `cross_val_predict` to get the predictions in the case above...

NicolasHug replied:
Yes I was confused by `df` too. We can rename it as `predictions` or just `preds`.

NicolasHug replied:
> Also I can't think of an elegant way to add the _reshape_df

Would it help to also add a `_get_method` helper, like

```py
def _get_method(clf):
	# return decision_function (the actual method) or predict_proba or raise error
```

?

lucyleeow replied:
Well funny you said that because I was just looking into using [`_check_classifier_response_method`](https://github.com/scikit-learn/scikit-learn/blob/6c0e04a7e3d156d1b4246b74b8c0fd66296fbd78/sklearn/metrics/_plot/base.py#L6) (from the plotting module) which does just that.
The thing is that inside `_get_predictions` we get the preds first, before we reshape it. Whereas above, we already have the preds and we just need to reshape. I could use `_check_classifier_response_method` to get the method, get the preds if necessary, then use the `reshape_perds` function to reshape. Is this too complicated?

NicolasHug replied:
I'm not sure about using ` _check_classifier_response_method ` because it gives higher priority to `predict_proba`.

One way to do it would be something like:

```py
def get_prediction_method(clf, return_string=False):
	# return clf.decision_function or clf.predict_proba or their corresponding string
	# we need strings when using cross_val_predict

def _get_predictions(X, pred_func):
	# pred_func is either a method of the clf, or a partial cross_val_predict
	preds = pred_func(X)
	# reshape preds here
	return preds

# and then we'd  call _get_predictions with one of:
_get_predictions(X, get_prediction_method(clf, return_string=False)  # when ensemble=True
_get_predictions(X, partial(cross_val_predict(..., method=get_prediction_method(clf, return_string=True)))  # when ensemble=False

```

If you think this looks too complex, let's just have a `_reshape_preds` function.

> The thing is that inside _get_predictions we (need to) get the preds first, before we reshape it. Whereas above, we already have the preds and we just need to reshape

I think that's fine. At execution this would indeed repeat the checks `if hasattr...` twice, but that's crazy cheap anyway and it would still make the code clearer IMO (unless there's something I don't see)

lucyleeow replied:
> I'm not sure about using `_check_classifier_response_method` because it gives higher priority to `predict_proba`

You're right, I thought you could pick priority.

> If you think this looks too complex, let's just have a _reshape_preds function.

I can see how that would work but it is a little complicated (for me, as I'm not familiar with `partial`). Is the advantage 'readability' ?

NicolasHug replied:
the advantage is no code duplication and succinctness. If it comes at the cost of being too hard to read / maintain, it's not worth it though. You'll be the judge.

partials are a classic functional-programming tool but I guess they're more of an intermediate / advanced concept. We do use them, mostly in the tests though

lucyleeow replied:
Looking into it, `_fit_calibrator` returns a `_CalibratedClassifierPipeline` instance for each cv fold. For `ensemble=False`, this won't work (for how `_fit_calibrator` works atm) - we want to fit once with all the predictions. We would need to take `_fit_calibrator` out of the cv for loop.

NicolasHug replied:
> Looking into it, _fit_calibrator returns a _CalibratedClassifierPipeline instance for each cv fold

Only when ensemble=True, right?

There's no CV loop when `ensemble=False`, the "loop" is in `cross_val_predict` and we only need to call `_fit_calibrator` once. Maybe I misunderstood?

lucyleeow replied:
This is difficult over text! I may be confused too. 

At the moment we have:
```Python
class CalibratedClassifierCV():

    def fit():
        if self.cv == "prefit":
            # `_fit_calibrator` using all X and y
            # append one `_CalibratedClassiferPipeline` instance to `self.calibrated_classifiers_`
        else:
            if ensemble:
                for train, test in cv.split(X, y):
                    # fit classifier
                    # call `_fit_calibrator`, producing a `_CalibratedClassiferPipeline` instance
                    # append `_CalibratedClassiferPipeline` instance to `self.calibrated_classifiers_`,
                    # ending up with `n_cv` `_CalibratedClassiferPipeline` instances
            else:
                # Use `cross_val_predict` to get all `preds`
                # call `_fit_calibrator` using `preds` and y
                # append one `_CalibratedClassiferPipeline` instance to `self.calibrated_classifiers_`
```
we want to get rid of the last `else` as there is duplication of code there. With the new `_get_prediction_method` we have:

```Python
class CalibratedClassifierCV():

    def fit():
        if self.cv == "prefit":
            # `_fit_calibrator` using all X and y
            # append one `_CalibratedClassiferPipeline` instance to `self.calibrated_classifiers_`
        else:
            for train, test in cv.split(X, y):
                # fit classifier
                # call `_fit_calibrator`, which calls `_get_predictions`.
                # In case of `ensemble=False`, we use use `partial(cross_val_predict())`
                # `_fit_calibrator` also fits calibrators and produces `_CalibratedClassiferPipeline` instance
                #  Only one _CalibratedClassiferPipeline` is required when `ensemble=False`
                # (i.e. shouldn't be in the for loop)
                # when `ensemble=True` `n_cv` `_CalibratedClassiferPipeline` instances are needed 
                # (i.e., should be in the for loop) 
```

NicolasHug replied:
Sorry if I wasn't clear! I don't think we can move away from the first version (with the last `else`). The duplication of code that I wanted to avoid was:

1. getting the right method

```py

                if hasattr(base_estimator, "decision_function"):
                    base_estimator_method = "decision_function"
                elif hasattr(base_estimator, "predict_proba"):
                    base_estimator_method = "predict_proba"
```

2. reshaping the predictions

```py
                if base_estimator_method == "decision_function":
                    if preds.ndim == 1:
                        preds = preds[:, np.newaxis]
                else:
                    if len(self.label_encoder_.classes_) == 2:
                        preds = preds[:, 1]
```

Right now, these 2 things happen both in `CalibratedClassifierCV.fit` (for ensemble=False) and in `_get_predictions` (for ensemble=True).

So to expand on my previous suggestion:


```py
def get_prediction_method(clf, return_string=False):
	# return clf.decision_function or clf.predict_proba or their corresponding string
	# we need strings when using cross_val_predict

def _get_predictions(X, pred_func):
	# pred_func is either a method of the clf, or a partial cross_val_predict
	preds = pred_func(X)
	# reshape preds here
	return preds


class CalibratedClassifierCV():

    def fit():
        if self.cv == "prefit":
            # `_fit_calibrator` using all X and y
            # append one `_CalibratedClassiferPipeline` instance to `self.calibrated_classifiers_`
        else:
            if ensemble:
                for train, test in cv.split(X, y):
                    # fit classifier on train
					preds = _get_predictions(test, get_prediction_method(clf, return_string=False)
                    # call `_fit_calibrator`, producing a `_CalibratedClassiferPipeline` instance
                    # append `_CalibratedClassiferPipeline` instance to `self.calibrated_classifiers_`,
                    # ending up with `n_cv` `_CalibratedClassiferPipeline` instances
            else:
				preds = _get_predictions(X, partial(cross_val_predict(..., method=get_prediction_method(clf, return_string=True)))
                # call `_fit_calibrator` using `preds` and y
                # append one `_CalibratedClassiferPipeline` instance to `self.calibrated_classifiers_`

```

I think we might need to modify `_fit_calibrator` to only accept `preds` and not call `_get_predictions`

lucyleeow replied:
FYI:
> Also is df = df[:, 1:] just equivalent to df = df[:, 1]?

it seems that it is not the same, you get 2d output with `df[:, 1:]`

NicolasHug reviewed (commented):
Thanks @lucyleeow , made a first pass.

I think the refactoring makes sense, WDYT?

I haven't looked in detail at the changes related to the label encoders though. Are these needed here or can they wait for another PR?

lucyleeow commented on sklearn/calibration.py:
```diff
@@ -178,53 +206,44 @@ def fit(self, X, y, sample_weight=None):
             with suppress(AttributeError):
                 self.n_features_in_ = base_estimator.n_features_in_
             self.classes_ = self.base_estimator.classes_
+            self.label_encoder_ = LabelEncoder().fit(self.classes_)
```
No not useful. I think I only made it an attribute due the way it worked before refactoring, as everything was a method.

lucyleeow commented on sklearn/calibration.py:
```diff
@@ -311,145 +375,159 @@ def _more_tags(self):
+    -------
+    pipeline : _CalibratedClassiferPipeline instance
+    """
+    Y = label_binarize(y, classes=label_encoder_.classes_)
```
Weirdly there is a function version of `LabelBinarizer`, [`label_binarize`](https://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.label_binarize.html), which I guess performs the `transform` method of `LabelBinarizer`

lucyleeow replied:
There is no equivalent function for `LabelEncoder`.

lucyleeow commented on sklearn/calibration.py:
```diff
@@ -311,145 +375,159 @@ def _more_tags(self):
             if n_classes == 2:
-                k += 1
-            proba[:, k] = calibrator.predict(this_df)
+                # When binary, proba of clf_fitted.classes_[1]
+                # output but `pos_class_indices` = 0
```
I amended to:
```
                # When binary, df consists of predictions for
                # clf_fitted.classes_[1] but `pos_class_indices` = 0
```
not sure if this is clearer

lucyleeow commented:
> I haven't looked in detail at the changes related to the label encoders though. Are these needed here or can they wait for another PR?

There wasn't a big change to this. Previously we passed the classes to `_CalibratedClassifier`:

https://github.com/scikit-learn/scikit-learn/blob/fd237278e895b42abe8d8d09105cbb82dc2cbba7/sklearn/calibration.py#L180-L181

and then instantiate a new `LabelEncoder` in `_CalibratedClassifier.fit`:

https://github.com/scikit-learn/scikit-learn/blob/fd237278e895b42abe8d8d09105cbb82dc2cbba7/sklearn/calibration.py#L325-L329

(note: `self.classes` is never `None` when using `CalibratedClassifierCV`)

I just changed it so we do not pass the classes but the fitted `LabelEncoder` instance. I don't think it is a big change but I can change it back.

lucyleeow committed:
rename df to preds

lucyleeow committed:
lint

lucyleeow committed:
lint

lucyleeow committed:
sep pred and fit calib

lucyleeow committed:
lint

lucyleeow commented:
I've make the changes. I like that `_get_predictions` and `_fit_calibrator` is separate now. Not sure if it's overall better, will sleep on it. Also, is it appropriate to use `method.__name__` is appropriate to use, to get the str of the method?

I do think that it is a bit of duplication to have both `_check_classifier_response_method` and `get_prediction_method`. I think a single utility fun like:
```
get_prediction_method(estimator, method, allow_alt=False):
    """Return prediction method of `estimator`.

    Parameters
    ----------------
    estimator : Estimator instance

    method : {'decision_function', 'predict_proba'}

    allow_alt : bool, default=False
        Whether to return the alternative method, if `method` does not exist for `estimator`.
```
would do the trick. But maybe that is a separate issue.

lucyleeow committed:
use partial

NicolasHug commented on sklearn/calibration.py:
```diff
@@ -236,28 +254,69 @@ def fit(self, X, y, sample_weight=None):
+                    base_estimator, X, y, cv=cv,
+                    method=get_prediction_method(base_estimator,
+                                                 return_string=True)))
+                preds = _get_predictions(
+                    pred_method, X, label_encoder_
+                )
```
This seems to fit in one line?

lucyleeow replied:
Its 83 for me. I assume we are still using 80 restriction..?

NicolasHug replied:
are you sure? It's 72 for me ^^

lucyleeow replied:
Whoops I was looking at the wrong line, sorry. You are right!

NicolasHug commented on sklearn/calibration.py:
```diff
@@ -236,28 +254,69 @@ def fit(self, X, y, sample_weight=None):
+            else:
+                pred_method = partial(cross_val_predict(
+                    base_estimator, X, y, cv=cv,
+                    method=get_prediction_method(base_estimator,
+                                                 return_string=True)))
```
Maybe using `.__name__` as you suggested is a better alternative: we can remove the `return_string` option and just use `__name__` when we need to.

NicolasHug commented on sklearn/calibration.py:
```diff
@@ -236,28 +254,69 @@ def fit(self, X, y, sample_weight=None):
+                    )
+                    self.calibrated_classifiers_.append(calibrated_classifier)
+            else:
+                pred_method = partial(cross_val_predict(
```
Let's move `this_estimator = clone(base_estimator)` just above and use `this_estimator` instead of `base_estimator` here

NicolasHug commented on sklearn/calibration.py:
```diff
@@ -311,145 +370,202 @@ def _more_tags(self):
 
-class _CalibratedClassifier:
-    """Probability calibration with isotonic regression or sigmoid.
+def get_prediction_method(clf_fitted, return_string=False):
```
Let's keep this explicitly private, even though the file is private as well.

NicolasHug replied:
Also as mentioned above let's remove `return_string` and always return the method object, since we can just rely on the `__name__ attribute.

NicolasHug commented on sklearn/calibration.py:
```diff
@@ -311,145 +370,202 @@ def _more_tags(self):
-    --------
-    CalibratedClassifierCV
+def _reshape_preds(preds, method, n_classes):
+    """Reshape predictions when classification binary.
```
missing "is"?

But also this seems to reshape regardless of whether the problem is binary or not?

NicolasHug replied:
also since this only seems to be used in _get_predictions maybe we can just declare it within `_get_predictions` as a subfunction, or just write the code "inline".

NicolasHug commented on sklearn/calibration.py:
```diff
@@ -311,145 +370,202 @@ def _more_tags(self):
 
-        return df, idx_pos_class
+def _get_predictions(pred_method, X, label_encoder_):
+    """Returns predictions for `X` and the index of classes present.
```
this does not seem to return "the index of classes present"

lucyleeow replied:
Whoops, thanks!

NicolasHug commented on sklearn/calibration.py:
```diff
@@ -89,6 +90,24 @@ class CalibratedClassifierCV(BaseEstimator, ClassifierMixin,
             ``cv`` default value if None changed from 3-fold to 5-fold.
 
+    ensemble : bool, default=True
+        Determines how the calibrator is fit, if `cv` is not `'prefit'`.
```
We prefer `fitted` to avoid any ambiguity

```suggestion
        Determines how the calibrator is fitted, if `cv` is not `'prefit'`.
```

(there are 2 other occurrences)

NicolasHug commented on sklearn/calibration.py:
```diff
@@ -311,145 +370,202 @@ def _more_tags(self):
+    """
+    Y = label_binarize(y, classes=label_encoder_.classes_)
+    pos_class_indices = label_encoder_.transform(clf_fitted.classes_)
+    calibrated_classifiers = []
```
I think this should just be called `calibrators`.

The calibrated classifiers are what this function returns.

lucyleeow replied:
good point

NicolasHug commented on sklearn/calibration.py:
```diff
@@ -311,145 +370,202 @@ def _more_tags(self):
-        return self
+    Parameters
+    ----------
+    clf_fitted : Estimator instance
+        Fitted classifier.
+
+    calibrators_fitted : List of fitted estimator instances
```
nit: I'd be fine calling these `clf` and `calibrators` since the doc is clear, but as you prefer.

NicolasHug commented on sklearn/calibration.py:
```diff
@@ -311,145 +370,202 @@ def _more_tags(self):
-            transform(self.base_estimator.classes_)
 
-        return df, idx_pos_class
+def _get_predictions(pred_method, X, label_encoder_):
```
Here and in other places:

- since it's not an attribute in this context, we should just name it `label_encoder_`
- however, since we only use its `classes_` attribute, we should just pass `classes_`. Or even better, `n_classes`. It makes it more explicit that what this function ultimately needs to know is the number of classes. How we get it should be an implementation detail.

lucyleeow replied:
Good point. I've amended for `_get_predictions`.

For `_fit_calibrator` and for `_CalibratedClassiferPipeline.predict_proba()` we use `label_encoder_.transform()` to get the indices of the classes present. I only input `label_encoder_.classes_` and instantiate a new `LabelEncoder` instance but I think passing `label_encoder_` is better? WDYT?

NicolasHug reviewed (commented):
Thanks @lucyleeow , made another pass.

I think the use of `partial` is appropriate and the refactoring looks good to me. We'll need tests for the new `ensemble` parameter ;)

> I do think that it is a bit of duplication to have both _check_classifier_response_method and get_prediction_method [...]

Yes we can consider having a unified utility, but indeed this would be better done in a separate PR.

lucyleeow committed:
suggestions, update docstring

lucyleeow committed:
merge master, amend for joblib

lucyleeow committed:
lint

lucyleeow committed:
Merge branch 'master' into calbclf_ensemble

lucyleeow committed:
add tests

lucyleeow committed:
lint

NicolasHug commented on sklearn/calibration.py:
```diff
@@ -377,145 +417,231 @@ def _more_tags(self):
-    See also
-    --------
-    CalibratedClassifierCV
+def _get_predictions(pred_method, method_name, X, n_classes):
```
can't we use `pred_method.__name__` instead of passing `method_name`?

lucyleeow replied:
You are too quick!
Annoyingly, with:
```
                pred_method = partial(
                    cross_val_predict, this_estimator, X, y, cv=cv,
                    method=method_name
                )
```
you can't get the parameters of the model from `pred_method`. I agree, this is not pretty. Any suggestions?

lucyleeow replied:
Oh wait I think I can use `signature`.

lucyleeow commented:
Thanks for the review @NicolasHug!

I have added tests for `ensemble`, by adding parametrize to appropriate tests. I have made other changes that aren't related and I can change if they should not be in this PR
* I have added parametrized `'method', ['sigmoid', 'isotonic']` for some tests when suitable. I think this makes debugging/maintenance easier
* Added some comments to make it easier to understand checks
* Amended comment wording in 2 places (I think), to clarify

Of note, I had to increase the sample size in `test_calibration_multiclass`, here:

 
For smaller sample sizes, `method='isotonic'` and `ensemble=False` resulted in very variable log-loss (tried with 10 different random state numbers) and the log-loss for the calibrated classifier was sometimes worse than for `RandomForestClassifier` (on average log-loss was slightly better for calibrated classifier). It was interesting to note that for these data and `RandomForestClassifier`, `ensemble=False` always resulted in worse log-loss cf `ensemble=True`, for both isotonic and sigmoid.

lucyleeow committed:
use kwarg in cross val predict

lucyleeow committed:
wip

lucyleeow committed:
kwarg cv

lucyleeow committed:
fix kwargs partial

lucyleeow committed:
use signature get pred

lucyleeow commented:
@NicolasHug, I would like to set `random_state` in `LinearSVC` in the test `test_calibration_multiclass`. Can I do it here or in another PR?

NicolasHug commented:
feel free to do it here though if the test is unstable w.r.t. random_state, we might need to investigate and try to make to more robust

lucyleeow commented:
Oh no it the results are not variable (i.e., the tests consistently pass) - I just wanted to set it for nitpicky reasons. I can leave as is.

NicolasHug commented:
ok, in general we do prefer to set random_state so go ahead if you want!

lucyleeow committed:
use signature get pred

lucyleeow committed:
add test ensemble

lucyleeow committed:
set rand state svc

lucyleeow committed:
update docs

lucyleeow commented:
Thanks for your help @NicolasHug! I've added tests and amended the doc so ready for review. Thanks!

NicolasHug commented on sklearn/tests/test_calibration.py:
```diff
@@ -338,42 +388,56 @@ def test_calibration_nan_imputer():
+            assert np.all(np.hstack([proba[:, :i],
+                                     proba[:, i + 1:]]))
+        else:
+            # Check `proba` all 1/n_classes
```
am I correct to assume this is because the regressor was trained on an array with only zeros as features?

lucyleeow replied:
I think it's because each sample is its own class - there are 10 samples and 10 classes. With `LeaveOneOut` cross-validation, the test set contains just one sample, and this sample belongs to a class not present in the training subset! So the prediction output is always the same value for every test subset (for `LinearSVC` at least).

I tried `cross_val_predict` with this data locally and I actually get the warning, which I guess is ignored when we use pytest (?):

> RuntimeWarning: Number of classes in training fold (9) does not match total number of classes (10). Results may not be appropriate for your use case. To fix this, use a cross-validation technique resulting in properly stratified folds

It's a bit odd, but I think each sample has to be it's own class for us to be able to test the results easily. Though this is confusing and I am not certain.

FYI It's related to this issue #7796 (fixed in #7799), where the probas weren't summing to 1 with `LeaveOneOut` because the test subset (1 sample) contained less classes than the train subset.

NicolasHug commented on doc/modules/calibration.rst:
```diff
@@ -107,22 +107,30 @@ Usage
+list. The output of :term:`predict` is the class that has the highest
+probability.
+
+When `ensemble=False`, cross-validation is used to obtain 'unbiased'
```
we should explicitly mention cross_cal_predict here.

NicolasHug reviewed (commented):
made another pass, thanks for your patience and persistence  @lucyleeow

lucyleeow commented on sklearn/calibration.py:
```diff
@@ -287,43 +319,47 @@ def fit(self, X, y, sample_weight=None):
+                method_name = _get_prediction_method(this_estimator).__name__
+                pred_method = partial(
+                    cross_val_predict, estimator=this_estimator, X=X, y=y,
+                    cv=cv, method=method_name
```
Will add all 3 here and have the removal dealt with in #18030 ?

lucyleeow commented on sklearn/calibration.py:
```diff
@@ -377,145 +413,229 @@ def _more_tags(self):
-    --------
-    CalibratedClassifierCV
+def _get_predictions(pred_method, X, n_classes):
+    """Returns predictions for `X` and reshapes binary outputs to shape
```
No problem, good to know!

lucyleeow commented on sklearn/calibration.py:
```diff
@@ -377,145 +413,229 @@ def _more_tags(self):
+        All the prediction classes.
+    """
+    def __init__(self, clf, calibrators, classes):
+        self.clf = clf
+        self.calibrators = calibrators
+        self.classes = classes
```
At the moment these are the classes:
https://github.com/scikit-learn/scikit-learn/blob/2f6e22b39841678286c6cc9c1842e6c695ed10b6/sklearn/calibration.py#L424-L427

So I will add `method`. Should I add `_` to the end of `classes` and `calibrators` ?

ogrisel replied:
The `_` is reserved for attributes that are computed when calling fit on an estimator class.

lucyleeow commented on sklearn/tests/test_calibration.py:
```diff
@@ -276,6 +298,33 @@ def test_calibration_prefit():
+
+    cal_clf = CalibratedClassifierCV(clf, method=method, cv=3, ensemble=False)
+    cal_clf.fit(X, y)
+    cal_probas = cal_clf.predict_proba(X[:10, :])
```
Good point!

lucyleeow commented on sklearn/tests/test_calibration.py:
```diff
@@ -338,42 +388,56 @@ def test_calibration_nan_imputer():
-    clf_prob = CalibratedClassifierCV(clf, method="sigmoid", cv=LeaveOneOut())
+    clf = LinearSVC(C=1.0, random_state=7)
+    clf_prob = CalibratedClassifierCV(
+        clf, method="sigmoid", cv=LeaveOneOut(), ensemble=ensemble
```
Good pick. I think this test should be broken up, as it tests different disjoint things. e.g., this just checks that only 'sigmoid' and 'isotonic' are allowed:
https://github.com/scikit-learn/scikit-learn/blob/2f6e22b39841678286c6cc9c1842e6c695ed10b6/sklearn/tests/test_calibration.py#L96-L98

I think breaking it up into self-contained tests that each check one specific thing helps maintenance. And with the parametrization, it avoids redundancy as some parts do not need the different `method` and `ensemble` options.

If you're happy I can break it up, or just add the parametrization?

lucyleeow replied:
Will break up the test but keep it in one commit so I can easily revert if not appropriate.

lucyleeow committed:
suggestions

lucyleeow committed:
param test_calib, break into 3 tests, add data fixture

lucyleeow committed:
lint

ogrisel reviewed (commented):
I put some comments on the refactoring itself on the parent PR: https://github.com/scikit-learn/scikit-learn/pull/17803#pullrequestreview-459156831

Here are comments that are more specific to the new ensemble hyperparam.

lucyleeow commented on sklearn/calibration.py:
```diff
@@ -377,145 +415,235 @@ def _more_tags(self):
-            Target values.
+    preds : array-like, shape (n_samples, n_classes) or (n_samples, 1) when \
+            binary.
+        Predictions for calibrating the predictions.
```
:rofl:

lucyleeow commented on sklearn/calibration.py:
```diff
@@ -377,145 +415,235 @@ def _more_tags(self):
-            transform(self.base_estimator.classes_)
 
-        return df, idx_pos_class
+def _fit_calibrator(clf, preds, y, classes, method, sample_weight=None):
```
Changed all `preds` to `predictions`

lucyleeow commented on sklearn/calibration.py:
```diff
@@ -377,145 +415,235 @@ def _more_tags(self):
-                                 '"isotonic". Got %s.' % self.method)
-            calibrator.fit(this_df, Y[:, k], sample_weight)
-            self.calibrators_.append(calibrator)
+class _CalibratedClassiferPipeline:
```
Yeah the pipeline can be a bit confusing, especially as we talk about classifier, calibrator 'pairs' in the doc and not 'pipeline'. Will change but @NicolasHug let me know if you feel differently.

lucyleeow committed:
og suggests

lucyleeow committed:
merge master

lucyleeow committed:
pred -> predictons

ogrisel committed:
Merge branch 'master' into calbclf_ensemble

lucyleeow commented on sklearn/calibration.py:
```diff
@@ -349,145 +386,244 @@ def _more_tags(self):
+    return calibrated_classifier
+
+
+def _get_prediction_method(clf):
```
Agreed. It was on my list too. To have something that would work for the plotting funs and here I suggested this https://github.com/scikit-learn/scikit-learn/pull/17856#issuecomment-656302895

lucyleeow committed:
suggestion

lucyleeow committed:
Merge branch 'calbclf_ensemble' of github.com:lucyleeow/scikit-learn into calbclf_ensemble

lucyleeow committed:
merge master

lucyleeow commented:
ping @NicolasHug @ogrisel 
It would be great to get this in. As there is such extensive change to `calibration.py`, merge conflicts are confusing to fix, e.g., cause by #18332

ogrisel commented on doc/modules/calibration.rst:
```diff
@@ -96,8 +96,8 @@ in [0, 1]. Denoting the output of the classifier for a given sample by :math:`f_
-samples used to fit the classifier, as this would
-introduce bias. The classifier performance on its training data would be
+samples used to fit the classifier, as this would introduce bias.
+This is because performance of the classifier on its training data would be
```
```suggestion
This is because performance of the the classifier on its training data would be
```

ogrisel reviewed (approved):
This looks very good to me. Thank you very much @lucyleeow for your patience.

Here are some fixes / small suggestions for further improvements but otherwise it seems ready for second review and merge.

Please add a new entry to `doc/whats_new/0.24.rst`.

lucyleeow commented on doc/modules/calibration.rst:
```diff
@@ -96,8 +96,8 @@ in [0, 1]. Denoting the output of the classifier for a given sample by :math:`f_
-samples used to fit the classifier, as this would
-introduce bias. The classifier performance on its training data would be
+samples used to fit the classifier, as this would introduce bias.
+This is because performance of the classifier on its training data would be
```
Wait why 2x 'the' ?

ogrisel replied:
weird .... let's ignore :)

lucyleeow commented on sklearn/calibration.py:
```diff
@@ -122,21 +106,49 @@ class CalibratedClassifierCV(ClassifierMixin,
+        :func:`~sklearn.model_selection.cross_val_predict`, which are then
+        used for calibration. At prediction time, the classifier used is the
+        `base_estimator` trained on all the data.
+        Note this method is implemented when `probabilities=True` for
+        :mod:`sklearn.svm` estimators.
```
Thanks, that is much clearer than what I wrote.

lucyleeow commented on sklearn/tests/test_calibration.py:
```diff
@@ -47,139 +57,156 @@ def test_calibration():
+            clf, method=method, cv=2, ensemble=ensemble
+        )
+        # Note that this fit overwrites the fit on the entire training
+        # set
```
I think it means that calling fit here will over-rid the fit that was done above:
https://github.com/scikit-learn/scikit-learn/blob/5654da026b7f1186b3a840d1cb140b6598b0af61/sklearn/tests/test_calibration.py#L47

but also happy to delete

ogrisel replied:
I see. It's fine.

lucyleeow committed:
suggestions, whats new

lucyleeow commented on doc/modules/calibration.rst:
```diff
@@ -107,22 +107,32 @@ Usage
+couple where the classifier is the `base_estimator` trained on all the data.
+In this case the output of :term:`predict_proba` for
+:class:`CalibratedClassifierCV` is the predicted probabilities obtained
+from the single `(classifier, calibrator)` couple.
```
Yes, good addition. I recall that `ensemble=True` greatly improved performance when you use isotonic and little data. We found this out when amending the test `test_calibration_multiclass`.

lucyleeow committed:
expand benefit

lucyleeow commented:
Thanks for the review @ogrisel, suggestions added!

lucyleeow committed:
formatting

ogrisel commented on doc/whats_new/v0.24.rst:
```diff
@@ -58,6 +58,14 @@ Changelog
 
+- |Enhancement| Refactor :class:`calibration.CalibratedClassifierCV` so the
+  chaining of the classifier to its corresponding calibrator and fitting of
+  the calibrator is now separate. Also add `ensemble` parameter, which
```
I think you can skip the mention of the internal refactoring and focus on the new `ensemble` parameters which is what is really of interest to our users.

ogrisel commented on doc/whats_new/v0.24.rst:
```diff
@@ -58,6 +58,14 @@ Changelog
+  chaining of the classifier to its corresponding calibrator and fitting of
+  the calibrator is now separate. Also add `ensemble` parameter, which
+  implements calibration via an ensemble of calibrators (current method) or
+  just one calibrator using all the data (method implemented in
```
```suggestion
  just one calibrator using all the data (similar to the built-in feature implemented in
```

ogrisel commented on doc/whats_new/v0.24.rst:
```diff
@@ -58,6 +58,14 @@ Changelog
+  chaining of the classifier to its corresponding calibrator and fitting of
+  the calibrator is now separate. Also add `ensemble` parameter, which
+  implements calibration via an ensemble of calibrators (current method) or
+  just one calibrator using all the data (method implemented in
+  :mod:`sklearn.svm` estimators with the `probabilities=True` parameter).
+  :pr:`17856` by :user:`Lucy Liu <lucyleeow>`.
```
```suggestion
  just one calibrator using all the data (similar to the built-in feature of
  :mod:`sklearn.svm` estimators with the `probabilities=True` parameter).
  :pr:`17856` by :user:`Lucy Liu <lucyleeow>`.
```

lucyleeow committed:
suggestion

NicolasHug commented on doc/whats_new/v0.24.rst:
```diff
@@ -58,6 +58,13 @@ Changelog
+  of calibration via an ensemble of calibrators (current method) or
+  just one calibrator using all the data (similar to the built-in feature of
+  :mod:`sklearn.svm` estimators with the `probabilities=True` parameter).
+  :pr:`17856` by :user:`Lucy Liu <lucyleeow>`.
```
I think we should also credit @aesuli for the original issue and PR

(sorry, I don't have time to make another real review at the moment)

lucyleeow replied:
Thanks I forgot!

lucyleeow committed:
add user

ogrisel committed:
Merge branch 'master' of github.com:scikit-learn/scikit-learn into calbclf_ensemble

glemaitre commented on sklearn/tests/test_calibration.py:
```diff
@@ -47,139 +57,156 @@ def test_calibration():
+    X, y = data
+    clf_base_regressor = \
+        CalibratedClassifierCV(RandomForestRegressor(), ensemble=ensemble)
+    assert_raises(RuntimeError, clf_base_regressor.fit, X, y)
```
We should use pytest.

glemaitre commented on sklearn/tests/test_calibration.py:
```diff
@@ -47,139 +57,156 @@ def test_calibration():
 
     calib_clf.fit(X, y)
-    assert len(calib_clf.calibrated_classifiers_) == splits
+    if ensemble:
+        assert len(calib_clf.calibrated_classifiers_) == splits
+    else:
+        assert len(calib_clf.calibrated_classifiers_) == 1
```
```suggestion
    expected_n_clf = splits if ensemble else 1:
    assert len(calib_clf.calibrated_classifiers_) == expected_n_clf
```

glemaitre commented on sklearn/tests/test_calibration.py:
```diff
@@ -192,27 +219,30 @@ def test_calibration_multiclass():
+    # Check that log-loss of calibrated classifier is smaller than
+    # log-loss obtained by naively turning OvR decision function to
+    # probabilities via softmax
+    def softmax(y_pred):
```
there is a function defines in `sklearn.utils.extmath` already. can we use it?

lucyleeow replied:
The `extmath` fun seems to normalise:
```Python
    if copy:
        X = np.copy(X)
    max_prob = np.max(X, axis=1).reshape((-1, 1))
    X -= max_prob
    np.exp(X, X)
    sum_prob = np.sum(X, axis=1).reshape((-1, 1))
    X /= sum_prob
    return X
```
which I don't think is the same as the softmax fun defined here? Regardless the test fails if I change it...

glemaitre commented on sklearn/tests/test_calibration.py:
```diff
@@ -192,27 +219,30 @@ def test_calibration_multiclass():
+    cal_clf.fit(X_train, y_train)
+    probas = cal_clf.predict_proba(X_test)
+    # Check probabilities sum to 1
+    assert_array_almost_equal(np.sum(probas, axis=1), np.ones(len(X_test)))
```
assert_allclose would be better

glemaitre reviewed (approved):
It looks good. Just a couple of nitpicks.

lucyleeow committed:
suggestions

ogrisel committed:
More stable multiclass test with Brier score

ogrisel committed:
Unused import

ogrisel commented:
Merged! Thank you very much @lucyleeow!

ogrisel commented on sklearn/tests/test_calibration.py:
```diff
@@ -66,7 +66,7 @@ def test_calibration(data, method, ensemble):
                                        sparse.csr_matrix(X_test))]:
         cal_clf = CalibratedClassifierCV(
-            clf, method=method, cv=2, ensemble=ensemble
+            clf, method=method, cv=5, ensemble=ensemble
```
I used `cv=5` whenever we assess the quality of the calibration because cv=2 trains models on much small dataset and can degrade the base model performance, hence make the test quite brittle.
